### PR TITLE
Fix test failures on MariaDB

### LINKED ISSUE #965: Test failures on MariaDB performance_schema: OFF
I would report as I got 2 test failures on my local environment.

I tried it on current master latest branch (4a20d58a5a5b8096da3a220a3994853ca26c79c7).
Do you know what's wrong on my environment?

Thanks!

OS: Fedora 27
MariaDB 10.2
Ruby: 2.5.1

```
$ cat /etc/fedora-release 
Fedora release 27 (Twenty Seven)

$ rpm -q mariadb-devel mariadb-server
mariadb-devel-10.2.12-5.fc27.x86_64
mariadb-server-10.2.12-5.fc27.x86_64
```

Start MariaDB daemon.

```
$ sudo systemctl start mariadb
```

```
$ which mysql
/usr/bin/mysql

$ rpm -qf /usr/bin/mysql
mariadb-10.2.12-5.fc27.x86_64

$ mysql -u root -e 'CREATE DATABASE /*M!50701 IF NOT EXISTS */ test'

$ which ruby
/usr/local/ruby-2.5.1/bin/ruby

$ ruby -v
ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux]

$ which bundle
/usr/local/ruby-2.5.1/bin/bundle

$ bundle -v
Bundler version 1.16.1

$ bundle install --path vendor/bundle --without benchmarks development

$ bundle list
Gems included by the bundle:
  * ast (2.4.0)
  * bundler (1.16.1)
  * diff-lcs (1.3)
  * eventmachine (1.2.5)
  * mysql2 (0.5.1)
  * parallel (1.12.1)
  * parser (2.5.1.0)
  * powerpack (0.1.1)
  * rainbow (2.2.2)
  * rake (10.4.2)
  * rake-compiler (1.0.4)
  * rspec (3.7.0)
  * rspec-core (3.7.1)
  * rspec-expectations (3.7.0)
  * rspec-mocks (3.7.0)
  * rspec-support (3.7.1)
  * rubocop (0.50.0)
  * ruby-progressbar (1.9.0)
  * unicode-display_width (1.3.2)
```

Run unit tests.

```
$ bundle exec rake spec
mkdir -p tmp/x86_64-linux/mysql2/2.5.1
cd tmp/x86_64-linux/mysql2/2.5.1
/usr/local/ruby-2.5.1/bin/ruby -I. ../../../../ext/mysql2/extconf.rb
checking for rb_absint_size()... yes
checking for rb_absint_singlebit_p()... yes
checking for rb_wait_for_single_fd()... yes
-----
Using mysql_config at /usr/bin/mysql_config
-----
checking for mysql.h... yes
checking for errmsg.h... yes
checking for SSL_MODE_DISABLED in mysql.h... no
checking for MYSQL_OPT_SSL_ENFORCE in mysql.h... yes
checking for MYSQL.net.vio in mysql.h... no
checking for MYSQL.net.pvio in mysql.h... yes
checking for MYSQL_ENABLE_CLEARTEXT_PLUGIN in mysql.h... yes
checking for SERVER_QUERY_NO_GOOD_INDEX_USED in mysql.h... yes
checking for SERVER_QUERY_NO_INDEX_USED in mysql.h... yes
checking for SERVER_QUERY_WAS_SLOW in mysql.h... yes
checking for MYSQL_OPTION_MULTI_STATEMENTS_ON in mysql.h... yes
checking for MYSQL_OPTION_MULTI_STATEMENTS_OFF in mysql.h... yes
checking for my_bool in mysql.h... yes
-----
Setting libpath to /usr/lib64/
-----
creating Makefile
cd -
cd tmp/x86_64-linux/mysql2/2.5.1
/usr/bin/gmake
compiling ../../../../ext/mysql2/client.c
../../../../ext/mysql2/client.c: In function ‘rb_set_ssl_mode_option’:
../../../../ext/mysql2/client.c:140:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
../../../../ext/mysql2/client.c: At top level:
cc1: warning: unrecognized command line option ‘-Wno-used-but-marked-unused’
cc1: warning: unrecognized command line option ‘-Wno-static-in-inline’
cc1: warning: unrecognized command line option ‘-Wno-reserved-id-macro’
cc1: warning: unrecognized command line option ‘-Wno-missing-variable-declarations’
cc1: warning: unrecognized command line option ‘-Wno-documentation-unknown-command’
cc1: warning: unrecognized command line option ‘-Wno-disabled-macro-expansion’
cc1: warning: unrecognized command line option ‘-Wno-covered-switch-default’
cc1: warning: unrecognized command line option ‘-Wno-conditional-uninitialized’
cc1: warning: unrecognized command line option ‘-Wno-self-assign’
cc1: warning: unrecognized command line option ‘-Wno-constant-logical-operand’
cc1: warning: unrecognized command line option ‘-Wno-parentheses-equality’
compiling ../../../../ext/mysql2/infile.c
compiling ../../../../ext/mysql2/mysql2_ext.c
compiling ../../../../ext/mysql2/result.c
compiling ../../../../ext/mysql2/statement.c
linking shared-object mysql2/mysql2.so
cd -
mkdir -p tmp/x86_64-linux/stage/lib/mysql2
cp CHANGELOG.md tmp/x86_64-linux/stage/CHANGELOG.md
cp LICENSE tmp/x86_64-linux/stage/LICENSE
cp README.md tmp/x86_64-linux/stage/README.md
mkdir -p tmp/x86_64-linux/stage/examples
cp examples/eventmachine.rb tmp/x86_64-linux/stage/examples/eventmachine.rb
cp examples/threaded.rb tmp/x86_64-linux/stage/examples/threaded.rb
mkdir -p tmp/x86_64-linux/stage/ext/mysql2
cp ext/mysql2/client.c tmp/x86_64-linux/stage/ext/mysql2/client.c
cp ext/mysql2/client.h tmp/x86_64-linux/stage/ext/mysql2/client.h
cp ext/mysql2/extconf.rb tmp/x86_64-linux/stage/ext/mysql2/extconf.rb
cp ext/mysql2/infile.c tmp/x86_64-linux/stage/ext/mysql2/infile.c
cp ext/mysql2/infile.h tmp/x86_64-linux/stage/ext/mysql2/infile.h
cp ext/mysql2/mysql2_ext.c tmp/x86_64-linux/stage/ext/mysql2/mysql2_ext.c
cp ext/mysql2/mysql2_ext.h tmp/x86_64-linux/stage/ext/mysql2/mysql2_ext.h
cp ext/mysql2/mysql_enc_name_to_ruby.h tmp/x86_64-linux/stage/ext/mysql2/mysql_enc_name_to_ruby.h
cp ext/mysql2/mysql_enc_to_ruby.h tmp/x86_64-linux/stage/ext/mysql2/mysql_enc_to_ruby.h
cp ext/mysql2/result.c tmp/x86_64-linux/stage/ext/mysql2/result.c
cp ext/mysql2/result.h tmp/x86_64-linux/stage/ext/mysql2/result.h
cp ext/mysql2/statement.c tmp/x86_64-linux/stage/ext/mysql2/statement.c
cp ext/mysql2/statement.h tmp/x86_64-linux/stage/ext/mysql2/statement.h
cp ext/mysql2/wait_for_single_fd.h tmp/x86_64-linux/stage/ext/mysql2/wait_for_single_fd.h
cp lib/mysql2.rb tmp/x86_64-linux/stage/lib/mysql2.rb
cp lib/mysql2/client.rb tmp/x86_64-linux/stage/lib/mysql2/client.rb
cp lib/mysql2/console.rb tmp/x86_64-linux/stage/lib/mysql2/console.rb
cp lib/mysql2/em.rb tmp/x86_64-linux/stage/lib/mysql2/em.rb
cp lib/mysql2/error.rb tmp/x86_64-linux/stage/lib/mysql2/error.rb
cp lib/mysql2/field.rb tmp/x86_64-linux/stage/lib/mysql2/field.rb
cp lib/mysql2/result.rb tmp/x86_64-linux/stage/lib/mysql2/result.rb
cp lib/mysql2/statement.rb tmp/x86_64-linux/stage/lib/mysql2/statement.rb
cp lib/mysql2/version.rb tmp/x86_64-linux/stage/lib/mysql2/version.rb
mkdir -p tmp/x86_64-linux/stage/spec
cp spec/configuration.yml.example tmp/x86_64-linux/stage/spec/configuration.yml.example
mkdir -p tmp/x86_64-linux/stage/spec/em
cp spec/em/em_spec.rb tmp/x86_64-linux/stage/spec/em/em_spec.rb
cp spec/my.cnf.example tmp/x86_64-linux/stage/spec/my.cnf.example
mkdir -p tmp/x86_64-linux/stage/spec/mysql2
cp spec/mysql2/client_spec.rb tmp/x86_64-linux/stage/spec/mysql2/client_spec.rb
cp spec/mysql2/error_spec.rb tmp/x86_64-linux/stage/spec/mysql2/error_spec.rb
cp spec/mysql2/result_spec.rb tmp/x86_64-linux/stage/spec/mysql2/result_spec.rb
cp spec/mysql2/statement_spec.rb tmp/x86_64-linux/stage/spec/mysql2/statement_spec.rb
cp spec/rcov.opts tmp/x86_64-linux/stage/spec/rcov.opts
cp spec/spec_helper.rb tmp/x86_64-linux/stage/spec/spec_helper.rb
mkdir -p tmp/x86_64-linux/stage/spec/ssl
cp spec/ssl/ca-cert.pem tmp/x86_64-linux/stage/spec/ssl/ca-cert.pem
cp spec/ssl/ca-key.pem tmp/x86_64-linux/stage/spec/ssl/ca-key.pem
cp spec/ssl/ca.cnf tmp/x86_64-linux/stage/spec/ssl/ca.cnf
cp spec/ssl/cert.cnf tmp/x86_64-linux/stage/spec/ssl/cert.cnf
cp spec/ssl/client-cert.pem tmp/x86_64-linux/stage/spec/ssl/client-cert.pem
cp spec/ssl/client-key.pem tmp/x86_64-linux/stage/spec/ssl/client-key.pem
cp spec/ssl/client-req.pem tmp/x86_64-linux/stage/spec/ssl/client-req.pem
cp spec/ssl/gen_certs.sh tmp/x86_64-linux/stage/spec/ssl/gen_certs.sh
cp spec/ssl/pkcs8-client-key.pem tmp/x86_64-linux/stage/spec/ssl/pkcs8-client-key.pem
cp spec/ssl/pkcs8-server-key.pem tmp/x86_64-linux/stage/spec/ssl/pkcs8-server-key.pem
cp spec/ssl/server-cert.pem tmp/x86_64-linux/stage/spec/ssl/server-cert.pem
cp spec/ssl/server-key.pem tmp/x86_64-linux/stage/spec/ssl/server-key.pem
cp spec/ssl/server-req.pem tmp/x86_64-linux/stage/spec/ssl/server-req.pem
cp spec/test_data tmp/x86_64-linux/stage/spec/test_data
mkdir -p tmp/x86_64-linux/stage/support
cp support/5072E1F5.asc tmp/x86_64-linux/stage/support/5072E1F5.asc
cp support/libmysql.def tmp/x86_64-linux/stage/support/libmysql.def
cp support/mysql_enc_to_ruby.rb tmp/x86_64-linux/stage/support/mysql_enc_to_ruby.rb
cp support/ruby_enc_to_mysql.rb tmp/x86_64-linux/stage/support/ruby_enc_to_mysql.rb
install -c tmp/x86_64-linux/mysql2/2.5.1/mysql2.so lib/mysql2/mysql2.so
cp tmp/x86_64-linux/mysql2/2.5.1/mysql2.so tmp/x86_64-linux/stage/lib/mysql2/mysql2.so
/usr/local/ruby-2.5.1/bin/ruby -I/home/jaruga/git/mysql2/vendor/bundle/ruby/2.5.0/gems/rspec-core-3.7.1/lib:/home/jaruga/git/mysql2/vendor/bundle/ruby/2.5.0/gems/rspec-support-3.7.1/lib /home/jaruga/git/mysql2/vendor/bundle/ruby/2.5.0/gems/rspec-core-3.7.1/exe/rspec --pattern spec/\*\*\{,/\*/\*\*\}/\*_spec.rb

Randomized with seed 54596
...
Failures:

  1) Mysql2::Client should set default program_name in connect_attrs
     Failure/Error: expect(result.first['attr_value']).to eq($PROGRAM_NAME)
     
     NoMethodError:
       undefined method `[]' for nil:NilClass
     # ./spec/mysql2/client_spec.rb:480:in `block (2 levels) in <top (required)>'

  2) Mysql2::Client should set custom connect_attrs
     Failure/Error: expect(results['program_name']).to eq('my_program_name')
     
       expected: "my_program_name"
            got: nil
     
       (compared using ==)
     # ./spec/mysql2/client_spec.rb:489:in `block (2 levels) in <top (required)>'

Finished in 10.35 seconds (files took 0.16545 seconds to load)
325 examples, 2 failures, 1 pending

Failed examples:

rspec ./spec/mysql2/client_spec.rb:474 # Mysql2::Client should set default program_name in connect_attrs
rspec ./spec/mysql2/client_spec.rb:483 # Mysql2::Client should set custom connect_attrs

Randomized with seed 54596
```

### LINKED ISSUE #1152: tests failing with mariadb10.5: Mysql2::Statement
```
Failures:

  1) Mysql2::Error encoding returns error messages as UTF-8 by default
     Failure/Error: expect(bad_err.message).to include("??}\u001F")
       expected "You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '??}\\001F' at line 1" to include "??}\u001F"
     # ./spec/mysql2/error_spec.rb:59:in `block (4 levels) in <top (required)>'
     # ./spec/spec_helper.rb:17:in `with_internal_encoding'
     # ./spec/mysql2/error_spec.rb:52:in `block (3 levels) in <top (required)>'

  2) Mysql2::Statement should create a statement
     Failure/Error: expect { statement = @client.prepare 'SELECT 1' }.to change(&method(:stmt_count)).by(1)
       expected result to have changed by 1, but was changed by 0
     # ./spec/mysql2/statement_spec.rb:18:in `block (2 levels) in <top (required)>'

  3) Mysql2::Statement close should free server resources
     Failure/Error: expect { stmt.close }.to change(&method(:stmt_count)).by(-1)
       expected result to have changed by -1, but was changed by 0
     # ./spec/mysql2/statement_spec.rb:702:in `block (3 levels) in <top (required)>'

Finished in 11.72 seconds (files took 0.21288 seconds to load)
322 examples, 3 failures

Failed examples:

rspec ./spec/mysql2/error_spec.rb:51 # Mysql2::Error encoding returns error messages as UTF-8 by default
rspec ./spec/mysql2/statement_spec.rb:16 # Mysql2::Statement should create a statement
rspec ./spec/mysql2/statement_spec.rb:699 # Mysql2::Statement close should free server resources
```
As those tests are old and simple I do not have a clue what is causing this.
Exactly same code builds / passes tests in our Fedora Stable (mariadb: 10.4.17).

The change may have been caused by update to openssl version: 1.1.1i
(According to one of our tools, which might be wrong.)
_ _ _ _

Ruby version: both 2.7 and 3.0 are failing
mariadb-server version: 10.5.8
gem version: 0.5.3

### PR DESCRIPTION
**Pull request by junaruga**
This PR fixes #965 and fixes #1152 test failures related to MariaDB.

Please check each commit's comment.
Thanks.

### CODE REVIEW

**Review by sodabrew — APPROVED** [review]
Looks excellent, thank you for running this down!

**Comment by sodabrew** [comment]
There are still some test failures after removing the last allow-fail, though?

**Comment by junaruga** [comment]
> There are still some test failures after removing the last allow-fail, though?

For cases I removed the allow-fail, there is no test failures any more.

But there are still roughly 3 test failures in other CI cases.

1.

```
- {os: ubuntu-20.04, ruby: 2.4, db: mariadb10.3, allow-failure: true} # => it will be fixed after the PR #1183 will be merged.
```

2.

```
# Allow failure due to Mysql2::Error: Unknown system variable 'session_track_system_variables'.
- {os: ubuntu-16.04, ruby: 2.4, db: mariadb10.0, allow-failure: true} # => we might be fix this case with the same way on the PR #1183.
```

3.

```
# Allow failure due to the issue #1165.
- {os: ubuntu-20.04, ruby: 2.4, db: mysql80, allow-failure: true}
- {os: macos-latest, ruby: 2.4, db: mysql, allow-failure: true}
```

**Comment by junaruga** [comment]
Travis mysql55 started error since yesterday due to the package repository not found.
https://travis-ci.org/github/brianmario/mysql2/jobs/767276558

```
E: Failed to fetch https://downloads.apache.org/cassandra/debian/dists/39x/main/binary-amd64/Packages  404  Not Found
E: Failed to fetch https://downloads.apache.org/cassandra/debian/dists/39x/main/binary-i386/Packages  404  Not Found
```

**Comment by junaruga** [comment]
5. macOS MariaDB case `Mysql2::EM::Client` fails. I think it rarely happens. I saw it in the past.
https://github.com/brianmario/mysql2/pull/1193/checks?check_run_id=2363155750#step:9:536

**Comment by junaruga** [comment]
> So, I can add the allow_failure to the macOS MariaDB case again if you like.

I rebased the last allow-fail's commit keeping allow_failure to the macOS MariaDB case.
I also added one more commit to add allow_failures to mysql55 case in Travis for now.

**Comment by junaruga** [comment]
> I also added one more commit to add allow_failures to mysql55 case in Travis for now.

I think an possible alternative of mysql55 could be mysql56 on xenial or trusty if the mysql55's repo issue will be not fixed.

http://repo.mysql.com/apt/ubuntu/dists/xenial/
http://repo.mysql.com/apt/ubuntu/dists/trusty/

**Comment by junaruga** [comment]
@sodabrew ping.

**Comment by junaruga** [comment]
Finally travis-ci.org shut down will occur on May 31, 2021. You can see https://github.com/brianmario/mysql2/issues/1149#issuecomment-835549157 for detail.

File: spec/mysql2/client_spec.rb [hunk 1]
```diff
@@ -488,6 +494,7 @@ def run_gc
   end
 
   it "should set custom connect_attrs" do
+    skip("DON'T WORRY, THIS TEST PASSES - but PERFORMNCE SCHEMA is not enabled in your MySQL daemon.") unless performance_schema_enabled
```
Comment (sodabrew) [inline spec/mysql2/client_spec.rb:497]:
```suggestion
    skip("DON'T WORRY, THIS TEST PASSES - but PERFORMANCE SCHEMA is not enabled in your MySQL daemon.") unless performance_schema_enabled
```

File: spec/mysql2/client_spec.rb [hunk 2]
```diff
@@ -479,6 +484,7 @@ def run_gc
   end
 
   it "should set default program_name in connect_attrs" do
+    skip("DON'T WORRY, THIS TEST PASSES - but PERFORMNCE SCHEMA is not enabled in your MySQL daemon.") unless performance_schema_enabled
```
Comment (sodabrew) [inline spec/mysql2/client_spec.rb:487]:
```suggestion
    skip("DON'T WORRY, THIS TEST PASSES - but PERFORMANCE SCHEMA is not enabled in your MySQL daemon.") unless performance_schema_enabled
```

**Comment by junaruga** [comment]
Thanks for finding my typos and the merging!

**Pull request merged**